sys/kern/signal.c

```diff
@@ -144,30 +144,28 @@ int do_sigprocmask(int how, const sigset_t *set, sigset_t *oset) {
  * These limitations (plus the fact that we currently have very little thread
  * states) make the logic of sending a signal very simple!
  */
-void sig_kill(proc_t *proc, signo_t sig) {
-  assert(proc != NULL);
+void sig_kill(proc_t *p, signo_t sig) {
+  assert(p != NULL);
   assert(mtx_owned(all_proc_mtx));
-  assert(mtx_owned(&proc->p_lock));
+  assert(mtx_owned(&p->p_lock));
   assert(sig < NSIG);
 
   /* Zombie processes shouldn't accept any signals. */
-  if (proc->p_state == PS_ZOMBIE)
+  if (p->p_state == PS_ZOMBIE)
     return;
 
-  thread_t *td = proc->p_thread;
+  thread_t *td = p->p_thread;
 
-  sig_t handler = proc->p_sigactions[sig].sa_handler;
+  sig_t handler = p->p_sigactions[sig].sa_handler;
   bool continued = sig == SIGCONT || sig == SIGKILL;
 
   /* If the signal is ignored, don't even bother posting it,
    * unless it's waking up a stopped process. */
-  if (proc->p_state == PS_STOPPED && continued) {
-    proc->p_state = PS_NORMAL;
-    proc->p_flags &= ~PF_STOPPED;
-    proc->p_flags |= PF_CONTINUED;
-    proc_t *parent = proc->p_parent;
-    if (parent)
-      cv_broadcast(&parent->p_waitcv);
+  if (p->p_state == PS_STOPPED && continued) {
+    p->p_state = PS_NORMAL;
+    p->p_flags &= ~PF_STOPPED;
+    p->p_flags |= PF_CONTINUED;
+    cv_broadcast(&p->p_parent->p_waitcv);
```

Comment:
Doesn't `init` have no parent (i.e. its `p_parent` is `NULL`)?

Reply:
I've been thinking of introducing `proc0` structure as a [sentinel value](https://en.wikipedia.org/wiki/Sentinel_value) to fix similar issues. I don't want to have extra checks scattered around the code to test for that very rare case where we handle `init` process.